File: sdks/node/conformance.js
import fs from "node:fs";
import { AppBridge } from "./src/index.js";

const manifest = JSON.parse(fs.readFileSync(new URL("../../protocol/conformance/app-schema.json", import.meta.url)));
const users = [];
const bridge = new AppBridge({ ...manifest, sdk: "@webtest/node", sdk_version: "0.1.0" });
bridge.register("create_user", ({ email, admin = false }) => {
  if (users.some((user) => user.email === email)) {
    throw Object.assign(new Error("a user with that email already exists"), { code: "user.email_taken" });
  }
  const user = { id: users.length + 1, email, admin };
  users.push(user);
  return user;
});
for (const name of Object.keys(manifest.functions).filter((name) => name.startsWith("echo_"))) {
  bridge.register(name, ({ value = null }) => value);
}
bridge.register("wait", ({ delay_ms }, { signal, emit }) => new Promise((resolve, reject) => {
  emit("progress", { phase: "waiting" });
  const timer = setTimeout(() => resolve("completed"), delay_ms);
  signal.addEventListener("abort", () => {
    clearTimeout(timer);
    reject(Object.assign(new Error("call was cancelled"), { code: "call.cancelled" }));
  }, { once: true });
}));
await bridge.connectFromEnv();

File: sdks/ruby/conformance.rb
require "json"
require_relative "lib/webtest"

manifest = JSON.parse(File.read(File.expand_path("../../protocol/conformance/app-schema.json", __dir__)))
manifest["sdk"] = "webtest-ruby"
manifest["sdk_version"] = "0.1.0"
users = []
bridge = WebTest::AppBridge.new(manifest)
bridge.register("create_user") do |arguments, _emit|
  email = arguments.fetch("email")
  if users.any? { |user| user["email"] == email }
    raise WebTest::ApplicationError.new("a user with that email already exists", code: "user.email_taken")
  end
  user = { "id" => users.length + 1, "email" => email, "admin" => arguments.fetch("admin", false) }
  users << user
  user
end
manifest.fetch("functions").each_key do |name|
  bridge.register(name) { |arguments, _emit| arguments.fetch("value", nil) } if name.start_with?("echo_")
end
bridge.register("wait") do |arguments, emit|
  emit.call("progress", { "phase" => "waiting" })
  sleep(arguments.fetch("delay_ms") / 1000.0)
  "completed"
end
bridge.connect_from_env
